sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -481,12 +481,12 @@ case class Add(
     // TODO: do not reorder consecutive `Add`s with different `evalMode`
     val reorderResult =
       orderCommutative({ case Add(l, r, _) => Seq(l, r) }).reduce(Add(_, _, evalMode))
-    if (resolved && reorderResult.resolved && reorderResult.dataType == dataType) {
-      reorderResult
+    if (resolved && reorderResult.resolved && reorderResult.dataType != dataType) {
+      // SPARK-40903: Append cast for the canonicalization of decimal Add if the result data type is
+      // changed. Otherwise, it may cause data checking error within ComplexTypeMergingExpression.
+      Cast(reorderResult, dataType)
```

Comment:
This seems to be the same idea that came up previously: https://github.com/apache/spark/pull/38379#discussion_r1004195334 but my concerns (https://github.com/apache/spark/pull/38379#discussion_r1004215562) were wrong, so LGTM.

Reply:
Yes, the cast is better after seconds thought.